{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Census datasets presence\n",
    "\n",
    "*Goal:* demonstrate basic use of the `datasets_presence_matrix` array.\n",
    "\n",
    "The presence matrix is a sparse array, indicating which features (var) were present in each dataset.  The array has dimensions [n_datasets, n_var], and is stored in the SOMA Measurement `varp` collection. The first dimension is indexed by the `soma_joinid` in the `census_datasets` dataframe. The second is indexed by the `soma_joinid` in the `var` dataframe of the measurement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>03f821b4-87be-4ff4-b65a-b5fc00061da7</td>\n",
       "      <td>Local and systemic responses to SARS-CoV-2 inf...</td>\n",
       "      <td>10.1038/s41586-021-04345-x</td>\n",
       "      <td>edc8d3fe-153c-4e3d-8be0-2108d30f8d70</td>\n",
       "      <td>Airway</td>\n",
       "      <td>edc8d3fe-153c-4e3d-8be0-2108d30f8d70.h5ad</td>\n",
       "      <td>236977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>03f821b4-87be-4ff4-b65a-b5fc00061da7</td>\n",
       "      <td>Local and systemic responses to SARS-CoV-2 inf...</td>\n",
       "      <td>10.1038/s41586-021-04345-x</td>\n",
       "      <td>2a498ace-872a-4935-984b-1afa70fd9886</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>2a498ace-872a-4935-984b-1afa70fd9886.h5ad</td>\n",
       "      <td>422220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>43d4bb39-21af-4d05-b973-4c1fed7b916c</td>\n",
       "      <td>Transcriptional Programming of Normal and Infl...</td>\n",
       "      <td>10.1016/j.celrep.2018.09.006</td>\n",
       "      <td>f512b8b6-369d-4a85-a695-116e0806857f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>f512b8b6-369d-4a85-a695-116e0806857f.h5ad</td>\n",
       "      <td>68036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0434a9d4-85fd-4554-b8e3-cf6c582bb2fa</td>\n",
       "      <td>Acute COVID-19 cohort across a range of WHO ca...</td>\n",
       "      <td>10.1101/2020.11.20.20227355</td>\n",
       "      <td>fa8605cf-f27e-44af-ac2a-476bee4410d3</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>fa8605cf-f27e-44af-ac2a-476bee4410d3.h5ad</td>\n",
       "      <td>59506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>3472f32d-4a33-48e2-aad5-666d4631bf4c</td>\n",
       "      <td>A single-cell transcriptome atlas of the adult...</td>\n",
       "      <td>10.15252/embj.2018100811</td>\n",
       "      <td>d5c67a4e-a8d9-456d-a273-fa01adb1b308</td>\n",
       "      <td>Retina</td>\n",
       "      <td>d5c67a4e-a8d9-456d-a273-fa01adb1b308.h5ad</td>\n",
       "      <td>19694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>347</td>\n",
       "      <td>f70ebd97-b3bc-44fe-849d-c18e08fe773d</td>\n",
       "      <td>A transcriptomic atlas of the mouse cerebellum...</td>\n",
       "      <td>10.1101/2020.03.04.976407</td>\n",
       "      <td>e0ed3c55-aff6-4bb7-b6ff-98a2d90b890c</td>\n",
       "      <td>A transcriptomic atlas of the mouse cerebellum</td>\n",
       "      <td>e0ed3c55-aff6-4bb7-b6ff-98a2d90b890c.h5ad</td>\n",
       "      <td>611034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>348</td>\n",
       "      <td>5d445965-6f1a-4b68-ba3a-b8f765155d3a</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>10.1038/s41586-020-2922-4</td>\n",
       "      <td>e04daea4-4412-45b5-989e-76a9be070a89</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>\n",
       "      <td>e04daea4-4412-45b5-989e-76a9be070a89.h5ad</td>\n",
       "      <td>9409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>349</td>\n",
       "      <td>5d445965-6f1a-4b68-ba3a-b8f765155d3a</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>10.1038/s41586-020-2922-4</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e.h5ad</td>\n",
       "      <td>65662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>350</td>\n",
       "      <td>17481d16-ee44-49e5-bcf0-28c0780d8c4a</td>\n",
       "      <td>Single-Cell Sequencing of Developing Human Gut...</td>\n",
       "      <td>10.1016/j.devcel.2020.11.010</td>\n",
       "      <td>8e47ed12-c658-4252-b126-381df8d52a3d</td>\n",
       "      <td>Paediatric Human Gut (4-14y)</td>\n",
       "      <td>8e47ed12-c658-4252-b126-381df8d52a3d.h5ad</td>\n",
       "      <td>22502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>351</td>\n",
       "      <td>17481d16-ee44-49e5-bcf0-28c0780d8c4a</td>\n",
       "      <td>Single-Cell Sequencing of Developing Human Gut...</td>\n",
       "      <td>10.1016/j.devcel.2020.11.010</td>\n",
       "      <td>b46237d1-19c6-4af2-9335-9854634bad16</td>\n",
       "      <td>Fetal Human Gut (6-11 PCW)</td>\n",
       "      <td>b46237d1-19c6-4af2-9335-9854634bad16.h5ad</td>\n",
       "      <td>62849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>352 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid                         collection_id  \\\n",
       "0              0  03f821b4-87be-4ff4-b65a-b5fc00061da7   \n",
       "1              1  03f821b4-87be-4ff4-b65a-b5fc00061da7   \n",
       "2              2  43d4bb39-21af-4d05-b973-4c1fed7b916c   \n",
       "3              3  0434a9d4-85fd-4554-b8e3-cf6c582bb2fa   \n",
       "4              4  3472f32d-4a33-48e2-aad5-666d4631bf4c   \n",
       "..           ...                                   ...   \n",
       "347          347  f70ebd97-b3bc-44fe-849d-c18e08fe773d   \n",
       "348          348  5d445965-6f1a-4b68-ba3a-b8f765155d3a   \n",
       "349          349  5d445965-6f1a-4b68-ba3a-b8f765155d3a   \n",
       "350          350  17481d16-ee44-49e5-bcf0-28c0780d8c4a   \n",
       "351          351  17481d16-ee44-49e5-bcf0-28c0780d8c4a   \n",
       "\n",
       "                                       collection_name  \\\n",
       "0    Local and systemic responses to SARS-CoV-2 inf...   \n",
       "1    Local and systemic responses to SARS-CoV-2 inf...   \n",
       "2    Transcriptional Programming of Normal and Infl...   \n",
       "3    Acute COVID-19 cohort across a range of WHO ca...   \n",
       "4    A single-cell transcriptome atlas of the adult...   \n",
       "..                                                 ...   \n",
       "347  A transcriptomic atlas of the mouse cerebellum...   \n",
       "348  A molecular cell atlas of the human lung from ...   \n",
       "349  A molecular cell atlas of the human lung from ...   \n",
       "350  Single-Cell Sequencing of Developing Human Gut...   \n",
       "351  Single-Cell Sequencing of Developing Human Gut...   \n",
       "\n",
       "                   collection_doi                            dataset_id  \\\n",
       "0      10.1038/s41586-021-04345-x  edc8d3fe-153c-4e3d-8be0-2108d30f8d70   \n",
       "1      10.1038/s41586-021-04345-x  2a498ace-872a-4935-984b-1afa70fd9886   \n",
       "2    10.1016/j.celrep.2018.09.006  f512b8b6-369d-4a85-a695-116e0806857f   \n",
       "3     10.1101/2020.11.20.20227355  fa8605cf-f27e-44af-ac2a-476bee4410d3   \n",
       "4        10.15252/embj.2018100811  d5c67a4e-a8d9-456d-a273-fa01adb1b308   \n",
       "..                            ...                                   ...   \n",
       "347     10.1101/2020.03.04.976407  e0ed3c55-aff6-4bb7-b6ff-98a2d90b890c   \n",
       "348     10.1038/s41586-020-2922-4  e04daea4-4412-45b5-989e-76a9be070a89   \n",
       "349     10.1038/s41586-020-2922-4  8c42cfd0-0b0a-46d5-910c-fc833d83c45e   \n",
       "350  10.1016/j.devcel.2020.11.010  8e47ed12-c658-4252-b126-381df8d52a3d   \n",
       "351  10.1016/j.devcel.2020.11.010  b46237d1-19c6-4af2-9335-9854634bad16   \n",
       "\n",
       "                                      dataset_title  \\\n",
       "0                                            Airway   \n",
       "1                                              PBMC   \n",
       "2                                              Skin   \n",
       "3                                             PBMCs   \n",
       "4                                            Retina   \n",
       "..                                              ...   \n",
       "347  A transcriptomic atlas of the mouse cerebellum   \n",
       "348   Krasnow Lab Human Lung Cell Atlas, Smart-seq2   \n",
       "349          Krasnow Lab Human Lung Cell Atlas, 10X   \n",
       "350                    Paediatric Human Gut (4-14y)   \n",
       "351                      Fetal Human Gut (6-11 PCW)   \n",
       "\n",
       "                             dataset_h5ad_path  dataset_total_cell_count  \n",
       "0    edc8d3fe-153c-4e3d-8be0-2108d30f8d70.h5ad                    236977  \n",
       "1    2a498ace-872a-4935-984b-1afa70fd9886.h5ad                    422220  \n",
       "2    f512b8b6-369d-4a85-a695-116e0806857f.h5ad                     68036  \n",
       "3    fa8605cf-f27e-44af-ac2a-476bee4410d3.h5ad                     59506  \n",
       "4    d5c67a4e-a8d9-456d-a273-fa01adb1b308.h5ad                     19694  \n",
       "..                                         ...                       ...  \n",
       "347  e0ed3c55-aff6-4bb7-b6ff-98a2d90b890c.h5ad                    611034  \n",
       "348  e04daea4-4412-45b5-989e-76a9be070a89.h5ad                      9409  \n",
       "349  8c42cfd0-0b0a-46d5-910c-fc833d83c45e.h5ad                     65662  \n",
       "350  8e47ed12-c658-4252-b126-381df8d52a3d.h5ad                     22502  \n",
       "351  b46237d1-19c6-4af2-9335-9854634bad16.h5ad                     62849  \n",
       "\n",
       "[352 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import cell_census\n",
    "\n",
    "census = cell_census.open_soma()\n",
    "\n",
    "# Grab the experiment containing human data, and the measurement therein with RNA\n",
    "human = census[\"census_data\"][\"homo_sapiens\"]\n",
    "human_rna = human.ms[\"RNA\"]\n",
    "\n",
    "# The cell census-wide datasets\n",
    "datasets_df = census[\"census_info\"][\"datasets\"].read_as_pandas_all()\n",
    "display(datasets_df)\n",
    "\n",
    "# The human RNA presence matrix\n",
    "presence = human.ms[\"RNA\"].varp[\"dataset_presence_matrix\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For convenience, read the entire presence matrix (for Homo sapiens) into a SciPy array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<352x60564 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "\twith 7220633 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the entire presence matrix. It may be returned in incremental chunks if larger than\n",
    "# read buffers, so concatenate into a single scipy.sparse.sp_matrix.\n",
    "\n",
    "# TODO: TileDB-SOMA#596 when implemented, will simplify this\n",
    "\n",
    "arrow_sparse_tensors = [t for t in presence.read_sparse_tensor((slice(None),))]\n",
    "flat_arrays = [t.to_numpy() for t in arrow_sparse_tensors]\n",
    "data = np.concatenate(tuple(t[0] for t in flat_arrays))\n",
    "coords = np.concatenate(tuple(t[1] for t in flat_arrays))\n",
    "presence_matrix = sparse.coo_matrix(\n",
    "    (data.flatten(), (coords.T[0].flatten(), coords.T[1].flatten())), shape=presence.shape\n",
    ").tocsr()\n",
    "presence_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also need the `var` dataframe, which is read into a Pandas DataFrame for convenient manipulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_df = human_rna.var.read_as_pandas_all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Is a feature present in a dataset?\n",
    "\n",
    "*Goal:* test if a given feature is present in a given dataset.\n",
    "\n",
    "**Important:** the presence matrix is indexed by soma_joinid, and is *NOT* positionally indexed.  In other words:\n",
    "* the first dimension of the presence matrix is the dataset's `soma_joinid`, as stored in the `census_datasets` dataframe.\n",
    "* the second dimension of the presence matrix is the feature's `soma_joinid`, as stored in the `var` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature is present.\n"
     ]
    }
   ],
   "source": [
    "var_joinid = var_df.loc[var_df.feature_id == \"ENSG00000286096\"].soma_joinid\n",
    "dataset_joinid = datasets_df.loc[datasets_df.dataset_id == \"97a17473-e2b1-4f31-a544-44a60773e2dd\"].soma_joinid\n",
    "is_present = presence_matrix[dataset_joinid, var_joinid][0, 0]\n",
    "print(f'Feature is {\"present\" if is_present else \"not present\"}.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What datasets contain a feature?\n",
    "\n",
    "*Goal:* look up all datasets that have a feature_id present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>304</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>a68b64d8-aee3-4947-81b7-36b8fe5a44d2</td>\n",
       "      <td>Tabula Sapiens - Stromal</td>\n",
       "      <td>a68b64d8-aee3-4947-81b7-36b8fe5a44d2.h5ad</td>\n",
       "      <td>82478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>305</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>97a17473-e2b1-4f31-a544-44a60773e2dd</td>\n",
       "      <td>Tabula Sapiens - Epithelial</td>\n",
       "      <td>97a17473-e2b1-4f31-a544-44a60773e2dd.h5ad</td>\n",
       "      <td>104148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>306</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>c5d88abe-f23a-45fa-a534-788985e93dad</td>\n",
       "      <td>Tabula Sapiens - Immune</td>\n",
       "      <td>c5d88abe-f23a-45fa-a534-788985e93dad.h5ad</td>\n",
       "      <td>264824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>307</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>5a11f879-d1ef-458a-910c-9b0bdfca5ebf</td>\n",
       "      <td>Tabula Sapiens - Endothelial</td>\n",
       "      <td>5a11f879-d1ef-458a-910c-9b0bdfca5ebf.h5ad</td>\n",
       "      <td>31691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>308</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
       "      <td>483152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid                         collection_id collection_name  \\\n",
       "304          304  e5f58829-1a66-40b5-a624-9046778e74f5  Tabula Sapiens   \n",
       "305          305  e5f58829-1a66-40b5-a624-9046778e74f5  Tabula Sapiens   \n",
       "306          306  e5f58829-1a66-40b5-a624-9046778e74f5  Tabula Sapiens   \n",
       "307          307  e5f58829-1a66-40b5-a624-9046778e74f5  Tabula Sapiens   \n",
       "308          308  e5f58829-1a66-40b5-a624-9046778e74f5  Tabula Sapiens   \n",
       "\n",
       "              collection_doi                            dataset_id  \\\n",
       "304  10.1126/science.abl4896  a68b64d8-aee3-4947-81b7-36b8fe5a44d2   \n",
       "305  10.1126/science.abl4896  97a17473-e2b1-4f31-a544-44a60773e2dd   \n",
       "306  10.1126/science.abl4896  c5d88abe-f23a-45fa-a534-788985e93dad   \n",
       "307  10.1126/science.abl4896  5a11f879-d1ef-458a-910c-9b0bdfca5ebf   \n",
       "308  10.1126/science.abl4896  53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
       "\n",
       "                    dataset_title                          dataset_h5ad_path  \\\n",
       "304      Tabula Sapiens - Stromal  a68b64d8-aee3-4947-81b7-36b8fe5a44d2.h5ad   \n",
       "305   Tabula Sapiens - Epithelial  97a17473-e2b1-4f31-a544-44a60773e2dd.h5ad   \n",
       "306       Tabula Sapiens - Immune  c5d88abe-f23a-45fa-a534-788985e93dad.h5ad   \n",
       "307  Tabula Sapiens - Endothelial  5a11f879-d1ef-458a-910c-9b0bdfca5ebf.h5ad   \n",
       "308    Tabula Sapiens - All Cells  53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad   \n",
       "\n",
       "     dataset_total_cell_count  \n",
       "304                     82478  \n",
       "305                    104148  \n",
       "306                    264824  \n",
       "307                     31691  \n",
       "308                    483152  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the feature's soma_joinid from the var dataframe\n",
    "var_joinid = var_df.loc[var_df.feature_id == \"ENSG00000286096\"].soma_joinid\n",
    "\n",
    "# The presence matrix is indexed by the joinids of the dataset and var dataframes,\n",
    "# so slice out the feature of interest by its joinid.\n",
    "dataset_joinids = presence_matrix[:, var_joinid].tocoo().row\n",
    "\n",
    "# From the datasets dataframe, slice out the datasets which have a joinid in the list\n",
    "datasets_df.loc[datasets_df.soma_joinid.isin(dataset_joinids)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What features are in a dataset?\n",
    "\n",
    "*Goal:* lookup the features present in a given dataset.\n",
    "\n",
    "This example also demonstrates the ability to do the query on multiple datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>3374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>9603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>6318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>2948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44644</th>\n",
       "      <td>44644</td>\n",
       "      <td>ENSG00000219926</td>\n",
       "      <td>OR7E104P</td>\n",
       "      <td>4672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44648</th>\n",
       "      <td>44648</td>\n",
       "      <td>ENSG00000267104</td>\n",
       "      <td>TBC1D3P1-DHX40P1</td>\n",
       "      <td>1841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44649</th>\n",
       "      <td>44649</td>\n",
       "      <td>ENSG00000265766</td>\n",
       "      <td>CXADRP3</td>\n",
       "      <td>1955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44651</th>\n",
       "      <td>44651</td>\n",
       "      <td>ENSG00000267453</td>\n",
       "      <td>CLEC4O</td>\n",
       "      <td>774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44654</th>\n",
       "      <td>44654</td>\n",
       "      <td>ENSG00000279274</td>\n",
       "      <td>RP11-533E23.2</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27211 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id      feature_name  feature_length\n",
       "0                0  ENSG00000121410              A1BG            3999\n",
       "1                1  ENSG00000268895          A1BG-AS1            3374\n",
       "2                2  ENSG00000148584              A1CF            9603\n",
       "3                3  ENSG00000175899               A2M            6318\n",
       "4                4  ENSG00000245105           A2M-AS1            2948\n",
       "...            ...              ...               ...             ...\n",
       "44644        44644  ENSG00000219926          OR7E104P            4672\n",
       "44648        44648  ENSG00000267104  TBC1D3P1-DHX40P1            1841\n",
       "44649        44649  ENSG00000265766           CXADRP3            1955\n",
       "44651        44651  ENSG00000267453            CLEC4O             774\n",
       "44654        44654  ENSG00000279274     RP11-533E23.2              75\n",
       "\n",
       "[27211 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Slice the dataset(s) of interest, and get the joinid(s)\n",
    "dataset_joinids = datasets_df.loc[datasets_df.collection_id == \"17481d16-ee44-49e5-bcf0-28c0780d8c4a\"].soma_joinid\n",
    "\n",
    "# Slice the presence matrix by the first dimension, i.e., by dataset\n",
    "var_joinids = presence_matrix[dataset_joinids, :].tocoo().col\n",
    "\n",
    "# From the feature (var) dataframe, slice out features which have a joinid in the list.\n",
    "var_df.loc[var_df.soma_joinid.isin(var_joinids)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.5 ('venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3da8ec1c162cd849e59e6ea2824b2e353dce799884e910aae99411be5277f953"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
